File: sdks/ios/Sources/MatrixMediaSDK/MMStream.swift
import Foundation
import Combine

/// An active stream connection.
///
/// Obtained via ``MMClient/joinStream(roomID:)`` or ``MMClient/startStream(roomID:config:)``.
/// Provides real-time state updates via published properties for SwiftUI integration.
///
/// ## SwiftUI Usage
/// ```swift
/// struct StreamView: View {
///     @ObservedObject var stream: MMStream
///
///     var body: some View {
///         VStack {
///             Text(stream.state == .connected ? "Live" : "Connecting...")
///             MMAudioRenderer(stream: stream)
///             Text("Viewers: \(stream.viewerCount)")
///             Button("Leave") {
///                 Task { await stream.leave() }
///             }
///         }
///     }
/// }
/// ```
@MainActor
public final class MMStream: ObservableObject {

    // MARK: - Public Published State

    /// Current connection state.
    @Published public private(set) var state: MMStreamState = .connecting

    /// The stream host participant, if known.
    @Published public private(set) var host: MMParticipant?

    /// Number of viewers currently connected.
    @Published public private(set) var viewerCount: Int = 0

    /// Current audio level (0.0 ... 1.0) for driving the ``MMAudioRenderer``.
    @Published public private(set) var audioLevel: Float = 0

    /// Whether the local microphone is muted (only meaningful for host streams).
    @Published public private(set) var isMuted: Bool = false

    /// Whether the stream currently has an active video track.
    @Published public private(set) var hasVideo: Bool = false

    /// Whether the active video track is a screen share (vs camera).
    @Published public private(set) var isScreenShare: Bool = false

    /// Whether end-to-end encryption is active on this stream.
    @Published public private(set) var e2eeEnabled: Bool = false

    /// The current E2EE key identifier, if E2EE is enabled.
    @Published public private(set) var e2eeKeyId: String?

    // MARK: - Public Read-Only

    /// The server-assigned stream identifier.
    public let streamID: String

    /// The Matrix room containing this stream.
    public let roomID: String

    // MARK: - Internal

    internal let bridge: LiveKitBridge
    private let apiClient: MMAPIClient
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Init

    init(
        streamID: String,
        roomID: String,
        bridge: LiveKitBridge,
        apiClient: MMAPIClient,
        e2ee: MME2eeInfo? = nil
    ) {
        self.streamID = streamID
        self.roomID = roomID
        self.bridge = bridge
        self.apiClient = apiClient

        if let e2ee = e2ee, e2ee.enabled {
            self.e2eeEnabled = true
            self.e2eeKeyId = e2ee.keyId
        }

        setupBridgeBindings()
    }

    // MARK: - Public API

    /// Leave the stream as a viewer, or stop publishing as a host.
    public func leave() async {
        state = .disconnected(reason: .userInitiated)
        await bridge.disconnect()

        // Notify the server (best-effort, don't throw on failure)
        try? await apiClient.leaveStream(streamID: streamID)
    }

    /// End the stream (host only). Disconnects all participants.
    ///
    /// - Throws: ``MMError/notAuthorized`` if the caller is not the host.
    public func stop() async throws {
        try await apiClient.endStream(streamID: streamID)
        state = .disconnected(reason: .streamEnded)
        await bridge.disconnect()
    }

    /// Mute or unmute the local microphone (host only).
    ///
    /// - Parameter muted: `true` to mute, `false` to unmute.
    public func setMuted(_ muted: Bool) {
        bridge.setMicrophoneEnabled(!muted)
        isMuted = muted
    }

    /// Set the playback volume for the remote audio.
    ///
    /// - Parameter volume: Volume from 0.0 (silent) to 1.0 (full).
    public func setVolume(_ volume: Float) {
        bridge.setVolume(max(0, min(1, volume)))
    }

    /// Enable the local camera and begin publishing video.
    ///
    /// - Throws: ``MMError/notAuthorized`` if the caller lacks publish permission,
    ///           or ``MMError/deviceUnavailable`` if no camera is available.
    public func enableCamera() async throws {
        try await bridge.enableCamera()
        hasVideo = true
        isScreenShare = false
    }

    /// Disable the local camera and stop publishing video.
    public func disableCamera() async {
        await bridge.disableCamera()
        hasVideo = false
    }

    /// Enable screen sharing and begin publishing the screen capture.
    ///
    /// - Throws: ``MMError/notAuthorized`` if the caller lacks publish permission.
    public func enableScreenShare() async throws {
        try await bridge.enableScreenShare()
        hasVideo = true
        isScreenShare = true
    }

    /// Disable screen sharing.
    public func disableScreenShare() async {
        await bridge.disableScreenShare()
        hasVideo = false
        isScreenShare = false
    }

    // MARK: - Bridge Bindings

    private func setupBridgeBindings() {
        // Bind bridge state changes to published properties.
        // The LiveKitBridge updates these via callbacks set here.

        bridge.onStateChange = { [weak self] newState in
            Task { @MainActor in
                self?.state = newState
            }
        }

        bridge.onAudioLevelChange = { [weak self] level in
            Task { @MainActor in
                self?.audioLevel = level
            }
        }

        bridge.onParticipantCountChange = { [weak self] count in
            Task { @MainActor in
                self?.viewerCount = count
            }
        }

        bridge.onHostUpdate = { [weak self] participant in
            Task { @MainActor in
                self?.host = participant
            }
        }

        bridge.onVideoTrackChange = { [weak self] hasVideo, isScreenShare in
            Task { @MainActor in
                self?.hasVideo = hasVideo
                self?.isScreenShare = isScreenShare
            }
        }
    }
}

File: sdks/android/mm-sdk/src/main/kotlin/com/matrixmedia/sdk/MMStream.kt
package com.matrixmedia.sdk

import com.matrixmedia.sdk.internal.livekit.LiveKitBridge
import com.matrixmedia.sdk.internal.networking.MMApiClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Represents an active MM stream session.
 *
 * Provides real-time state observation via [StateFlow] properties and
 * control methods for leaving, stopping, muting, and adjusting volume.
 *
 * Obtained from [MMClient.joinStream] (viewer) or [MMClient.startStream] (host).
 *
 * ## Lifecycle
 *
 * - Viewer: call [leave] to disconnect gracefully.
 * - Host: call [stop] to end the stream for all participants.
 *
 * The [state] flow emits [MMStreamState] values reflecting the connection
 * lifecycle, including automatic reconnection attempts on network changes.
 */
class MMStream internal constructor(
    private val bridge: LiveKitBridge,
    private val apiClient: MMApiClient,
    private val streamId: String,
    e2ee: MME2eeInfo? = null
) {
    private val _e2eeEnabled = MutableStateFlow(e2ee?.enabled == true)
    private val _e2eeKeyId = MutableStateFlow<String?>(e2ee?.takeIf { it.enabled }?.keyId)

    /** Whether end-to-end encryption is active on this stream. */
    val e2eeEnabled: StateFlow<Boolean> = _e2eeEnabled.asStateFlow()

    /** The current E2EE key identifier, if E2EE is enabled. */
    val e2eeKeyId: StateFlow<String?> = _e2eeKeyId.asStateFlow()

    /** Current stream connection state. */
    val state: StateFlow<MMStreamState>
        get() = bridge.streamState

    /** The current host participant, or null if unknown. */
    val host: StateFlow<MMParticipant?>
        get() = bridge.host

    /** Current number of viewers in the stream. */
    val viewerCount: StateFlow<Int>
        get() = bridge.viewerCount

    /**
     * Current audio level (0.0 = silence, 1.0 = maximum).
     *
     * Updated from LiveKit's built-in audio level callbacks. Useful for
     * driving [com.matrixmedia.sdk.ui.MMAudioRenderer] visualizations.
     */
    val audioLevel: StateFlow<Float>
        get() = bridge.audioLevel

    /** Whether the local audio output is muted. */
    val isMuted: StateFlow<Boolean>
        get() = bridge.isMuted

    /** Whether the stream currently has an active video track. */
    val hasVideo: StateFlow<Boolean>
        get() = bridge.hasVideo

    /** Whether the active video track is a screen share (vs camera). */
    val isScreenShare: StateFlow<Boolean>
        get() = bridge.isScreenShare

    /**
     * Leave the stream as a viewer.
     *
     * Notifies the MM server and disconnects from the SFU. After calling
     * this method, the [MMStream] instance is no longer usable.
     */
    suspend fun leave() {
        try {
            apiClient.leaveStream(streamId)
        } finally {
            bridge.disconnect()
        }
    }

    /**
     * Stop the stream (host only).
     *
     * Ends the stream for all participants. After calling this method,
     * the [MMStream] instance is no longer usable.
     *
     * @throws MMException.NotAuthorized if the caller is not the host.
     */
    suspend fun stop() {
        try {
            apiClient.endStream(streamId)
        } finally {
            bridge.disconnect()
        }
    }

    /**
     * Mute or unmute the audio output.
     *
     * For viewers, this mutes the incoming host audio. For hosts, this
     * mutes the outgoing microphone (other participants hear silence).
     *
     * @param muted `true` to mute, `false` to unmute.
     */
    fun setMuted(muted: Boolean) {
        bridge.setMuted(muted)
    }

    /**
     * Set the playback volume.
     *
     * @param volume Volume level from 0.0 (silent) to 1.0 (full volume).
     */
    fun setVolume(volume: Float) {
        require(volume in 0.0f..1.0f) { "Volume must be between 0.0 and 1.0, got $volume" }
        bridge.setVolume(volume)
    }

    /**
     * Enable the local camera and begin publishing video (host only).
     *
     * @throws MMException.NotAuthorized if the caller lacks publish permission.
     * @throws MMException.DeviceUnavailable if no camera is available.
     */
    suspend fun enableCamera() {
        bridge.enableCamera()
    }

    /**
     * Disable the local camera and stop publishing video.
     */
    suspend fun disableCamera() {
        bridge.disableCamera()
    }

    /**
     * Enable screen sharing and begin publishing the screen capture (host only).
     *
     * @throws MMException.NotAuthorized if the caller lacks publish permission.
     */
    suspend fun enableScreenShare() {
        bridge.enableScreenShare()
    }

    /**
     * Disable screen sharing.
     */
    suspend fun disableScreenShare() {
        bridge.disableScreenShare()
    }
}
